### PR TITLE
Support Task Metadata Endpoint v3

### PR DESCRIPTION
**Pull request by hayajo**
Supported [Task Metadata Endpoint v3](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v3.html).

- ecs-agent >= 1.26.1
    - https://github.com/aws/amazon-ecs-agent/pull/1932
- Add new 'ecs_v3' platform
- Use Task Metadata Endpoint v3 to retrieve metrics and spec
- Automatically resolves the startup type ('ec2' or 'fargate')
- Automatically detect the network types ('bride', 'host' or 'awsvpc')
- Only spec of agent's network interface is retrieved in 'EC2/bridge'
    - Because IP address and interface name are not associated

### CODE REVIEW

**Comment by astj** [comment]
> Automatically resolves the startup type ('ec2' or 'fargate')

Does Fargate supports task metadata endpoint v3? According to AWS doc Fargate seems to support only v2 currently, but I haven't check actual behavior yet.

File: agent/platform.go [hunk 1]
```diff
@@ -29,6 +30,11 @@ func NewPlatform(ctx context.Context, ignoreContainer *regexp.Regexp) (platform.
 	case platform.ECSAwsvpc:
 		return ecsawsvpc.NewECSAwsvpcPlatform(false, ignoreContainer)
 
+	case platform.ECSv3:
+		metadataURI := os.Getenv("ECS_CONTAINER_METADATA_URI")
```
Comment (astj) [inline agent/platform.go]: What happens if users mistakingly specify ECSv3 at unsupported environments like old container instances (or Fargate)?
Maybe it doesn't cause disaster 💥 with current implementation, but users will be happy if they can get feedback to know the container-agent is in unsupported environments.

File: provider/types.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package provider
```
Comment (astj) [inline provider/types.go]: Currently provider is used only inside platform/ecsv3, therefore it seems we don't have to place top-level of container-agent.
Is there any future plan to use this provider elsewhere?

Comment (hayajo) [reply provider/types.go]: It was also planned to be used on the Kubernetes platform, but it seems that just the platform package is enough.
Move to the ecsv3 package.

File: agent/platform.go (hunk 1 above)
Comment (hayajo) [reply agent/platform.go]: In the current implementation, the connection error to the endpoint is output in the log and the agent terminates.
Fix to give the user appropriate feedback.

File: platform/ecsv3/ecs.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	meta, err := p.client.GetTaskMetadata(ctx)
+	if err != nil {
+		logger.Warningf("failed to get metadata: %s", err)
+		return false
+	}
+	return isRunning(meta.KnownStatus)
+}
+
+func isRunning(status string) bool {
+	return strings.EqualFold("running", status)
```
[the remark this answers is not in the document]
Comment (hayajo) [reply platform/ecsv3/ecs.go]: "RUNNING" is assumed.
https://github.com/aws/amazon-ecs-agent/blob/aabe65ee15a013a6bc594da52b94d817419ee74a/agent/api/container/status/containerstatus.go#L62

I will correct that.

File: platform/ecsv3/ecs.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+func resolveProvider(executionEnv string) (provider, error) {
+	switch executionEnv {
+	case executionEnvFargate:
+		return fargateProvider, nil
+	case executionEnvEC2:
+		return ecsProvider, nil
+	default:
+		return provider("UNKNOWN"), errors.New("unknown exectution env")
```
[the remark this answers is not in the document]
Comment (hayajo) [reply platform/ecsv3/ecs.go]: Thank you!

**Comment by hayajo** [comment]
Thank you for the review.

**Pull request merged**